Fix broker docker build so new images compatible with main branch can be pushed.

timoxley opened the pull request:
* CI is failing for all of the cross-client tests that involve the JS client because
* the network nodes are incompatible with the trackers run in docker because
* the Docker images for the broker haven't been updated in a month because
* the broker's CI docker image build fails because
* the client's `sqlite3` dependency is trying to run an npm `install` script during the `npm prune --production`, and that `install` script uses a development dependency `node-pre-gyp` which was removed during that same `npm prune --production`.

😵‍💫 

Not clear how/why this changed, possibly a change in the npm 6 line, or a dependency of npm 6?

### Changes

* Set the Dockerfile to use `--ignore-scripts` with `npm prune`. This seems to fix the issue.
* Set the broker CI config to build & tests images even if they're not tagged or on main branch. i.e. it should build & test the docker image for PRs now. This allows me to test the above change and should help catch a broken docker push step sooner in the future.

It's possible this issue goes away with the move to npm 8 and npm workspaces in #259. The change to run the docker build step for PRs will help.

timoxley committed:
ci(broker): Build images for non-main branch.

timoxley committed:
build(broker): Ignore scripts during npm prune in Dockerfile.

timoxley committed:
build(broker): Run npm prune step on separate docker layer.

So previous, less error-prone steps can be cached.

timoxley committed:
ci(broker): Separate build+test & build+push jobs. Means build+test doesn't need to wait for other tests.

timoxley committed:
ci: Upload docker log artifacts on failure in broker & cross-client.

Remember to kill docker stack *after* collecting docker stack logs.

kare reviewed (approved):
LGTM

timoxley committed:
build: Split Docker build stage into multiple layers for convenience of caching.

timoxley committed:
build: Try improve Dockerfile npm speed with npm@8 --no-audit --prefer-offline, pruning in parallel, only pruning broker + deps.

kare commented on Dockerfile:
```diff
@@ -1,24 +1,23 @@
-RUN apt-get update && apt-get install --assume-yes --no-install-recommends curl \
-	&& apt-get clean \
-	&& rm -rf /var/lib/apt/lists/*
+FROM node:16-alpine
```
Please use Debian based images for Docker containers. Currently `buster` or `buster-slim`.

Reasoning is to minimise download times when pulling containers from Hub. This is achieved by sharing Docker container layers. Currently we aim to share the base layer and maybe later we can share more.

Following command creates a single layer that can be shared by other containers:
```
RUN apt-get install -y curl
```

timoxley replied:
Ah good point, though the reason I switched to alpine is the image is only ~5MB vs ~25MB for buster-slim.

kare replied:
Given that all Docker containers use Debian Buster base image I don't think the size of the base image is significant.

timoxley replied:
Changed.

kare commented on Dockerfile:
```diff
@@ -31,4 +30,6 @@ EXPOSE 7170/tcp
-CMD ./bin/broker.js # start broker from default config
+
+# start broker from default config
+CMD ["./bin/broker.js"]
```
```suggestion
ENTRYPOINT ["./bin/broker.js"]
```

I didn't try, but I think this should fix the issue where `init` is required.

timoxley replied:
@kare good thought, though I believe this will break the ability to use this image to start other executables

e.g. trackers in the docker-compose https://github.com/streamr-dev/streamr-docker-dev/blob/3c55bab982ed65d30722e65b36a0b0649298f0ce/docker-compose.yml#L128

something like this in the testnet instructions:
```
docker run -it -v $(cd ~/.streamrDocker; pwd):/root/.streamr \
  streamr/broker-node:testnet bin/config-wizard

```

timoxley committed:
ci(broker): Push broker image with 'debug' tag if docker test steps fail.

Allows debugging locally.
Failed builds use the 'debug' tag instead of 'dev'.
Only pushes debug/dev tags on main/tagged pushes.
Start nginx in test-docker job.

kare reviewed (approved):
LGTM
